### PR TITLE
Categorize imported transcript transfers

### LINKED ISSUE #685: Update Transcript Imports to automatically categorize non-UCI credits
When the user goes to import a transcript, it will currently throw everything into "Other Transfers", which is not ideal. We should make it so that the user's data automatically gets sorted into Courses & AP Exams whenever possible.

This will require logic used in #666 to automatically detect AP Exams, and should use a single batch course request for all "non-'AP '-prefixed items" to automatically detect courses transferred

Transcript parsing logic is currently on the frontend and does not need to be changed.

Instead, the transfers array you receive after parsing should simply be input to a new backend route that does AP Exam and Course detection. That backend route should return an array of UCI courses and an array of AP Exams cleared (should be the same JSON structure as used in Redesigned Transfers).

After receiving the response, it should be joined with what's currently on the frontend. Unique new items should be added to the frontend arrays in the `transferCreditsSlice` AND should be saved immediately (both to localStorage and to account – wait for localStorage PR to be merged). The backend route may need to be modified to do nothing on row insert conflict in case the user runs this in two different tabs.

### PR DESCRIPTION
**Pull request by CadenLee2**
Transfers are now properly categorized and saved when importing a transcript

## Description

- Updated the transfer conversion helper code to support optional scores as well
- Refactored the transfer conversion helper code somewhat (could use some advice on other steps to make it more readable)
- Refactored `userAPExam` to be a shared type across backend and frontend to reduce duplicate code (update  #693 with this as well, depending on which gets merged first)
- Transfers are now categorized (and saved to the db, if logged in) in the transfer import
  - Specifically, they are upserted (on the frontend, so that we are free to change `overrideAllTransfers` to be more accurate to its name) before being saved
- Note: we may want to do more testing with multiple tabs/handling collisions on the backend as mentioned in #685

## Screenshots

Since @fiveminus1 included a video I had to as well

ezgif-75fb02f721921e

## Test Plan

- [ ] Test with various transcripts (such as your own)
- [ ] Test while logged in and logged out, and saving to the account as well as locally (try reloading)
- [ ] Test if you currently have various exams, courses, etc. to ensure things don't get duplicated or erased
- [ ] Ensure scores are accurate, edge cases are handled (like multiple Calc ABs), etc.

## Issues

Closes #685

### CODE REVIEW

File: api/src/helpers/transferCredits.ts [hunk 1]
```diff
@@ -186,36 +189,30 @@ const handleApCalcAB = (
...
+const organizeApExam = (
+  transfer: GroupedUncategorizedTransfer,
   toInsertAp: TransferredApExamRow[],
   hasSubscore: boolean,
   allAps: ApExamBasicInfo[],
 ): boolean => {
-  const bestMatch = tryMatchAp(transferName, allAps);
+  const bestMatch = tryMatchAp(transfer.courseName, allAps);
   if (!bestMatch) {
     // Could not match; leave it here
```
Comment (Awesome-E) [inline api/src/helpers/transferCredits.ts:215]: "Leave it here" is a reference to the misc table, which is not necessary in this case.

```suggestion
  if (!bestMatch) return false;
```

File: api/src/helpers/transferCredits.ts [hunk 2]
```diff
@@ -228,26 +225,21 @@ export const organizeApExam = (
...
-  transfer: TransferredMiscSelectedRow,
-  transferName: string,
+  transfer: GroupedUncategorizedTransfer,
   toInsertCourse: TransferredCourseRow[],
   validCourses: Record<string, string>,
 ): Promise<boolean> => {
-  const bestMatch = tryMatchCourse(transferName, validCourses);
+  const bestMatch = tryMatchCourse(transfer.courseName, validCourses);
   if (!bestMatch) {
     // Could not match; leave it here
```
Comment (Awesome-E) [inline api/src/helpers/transferCredits.ts:242]: same as previous

File: api/src/helpers/transferCredits.ts [hunk 3]
```diff
@@ -17,10 +19,11 @@ type TransferredMiscRow = {
   units: number | null;
 };
 
-type TransferredMiscSelectedRow = {
+type GroupedUncategorizedTransfer = {
```
Comment (Awesome-E) [inline api/src/helpers/transferCredits.ts:22]: Consider adding a docstring to this type because "grouped" is not inherently clear on its meaning

File: api/src/helpers/transferCredits.ts [hunk 4]
```diff
@@ -324,33 +314,30 @@ const mergeDuplicateCourseResults = (toInsertCourse: TransferredCourseRow[]) =>
...
       existing.count++;
       existing.totalUnits += row.units ?? 0;
     } else {
-      lookup[row.name] = { totalUnits: row.units ?? 0, count: 1 };
+      lookup[row.name] = { totalUnits: row.units ?? 0, count: 1, score: row.score ?? undefined };
     }
   });
-
-  const uniqueRows: TransferredMiscSelectedRow[] = Object.entries(lookup).map(([courseName, data]) => ({
+  const uniqueRows: GroupedUncategorizedTransfer[] = Object.entries(lookup).map(([courseName, data]) => ({
```
Comment (Awesome-E) [inline api/src/helpers/transferCredits.ts:332]: If we're refactoring this whole file, I'd argue that these should be pulled out to helper functions so the main function is easier to read

File: api/src/helpers/transferCredits.ts [hunk 5]
```diff
@@ -324,33 +314,30 @@ const mergeDuplicateCourseResults = (toInsertCourse: TransferredCourseRow[]) =>
...
-  const transfers: TransferredMiscSelectedRow[] = uniqueRows
-    //.limit(40) // For testing, only look at a few entries
-    .filter((transfer) => transfer.courseName != null && transfer.count != 0) as TransferredMiscSelectedRow[];
+  // Obtain/clean up all relevant transfers (invalid ones that cannot be categorized are filtered out here)
+  const transfers: GroupedUncategorizedTransfer[] = uniqueRows
+    .filter((transfer) => transfer.courseName != null && transfer.count != 0)
+    .map((transfer) => ({ ...transfer, courseName: transfer.courseName.trim() })) as GroupedUncategorizedTransfer[];
 
   // Get all the users who explicitly have an AB subscore
   let hasSubscore = false;
```
Comment (Awesome-E) [inline api/src/helpers/transferCredits.ts:343]: Even the for loop here can be a helper function

something like
`const hasABSubscore = checkABSubscore(transfers, allAps)`

File: api/src/helpers/transferCredits.ts [hunk 6]
```diff
@@ -362,44 +349,27 @@ export const organizeLegacyTransfers = async (rows: TransferData[]) => {
...
-    const transferName = transfer.courseName.trim();
-    if (transferName.startsWith('AP ')) {
+    if (transfer.courseName.startsWith('AP ')) {
       continue;
     }
-    coursesToValidate.add(removeAllSpaces(transferName));
+    coursesToValidate.add(removeAllSpaces(transfer.courseName));
   }
-  // console.log('Validating courses from the API (batch)');
   const validCourses = await validateCoursesAPI(coursesToValidate);
```
Comment (Awesome-E) [inline api/src/helpers/transferCredits.ts:357]: similar idea here
`const validCourses = getValidCourseNames(transfers)`

would call a helper that builds the courses to validate set and loops through all transfers

File: api/src/helpers/transferCredits.ts [hunk 7]
```diff
@@ -362,44 +349,27 @@ export const organizeLegacyTransfers = async (rows: TransferData[]) => {
...
-    if (transferName.startsWith('AP ')) {
-      const reorganized = organizeApExam(
-        transfer as TransferredMiscSelectedRow,
-        transferName,
-        toInsertAp,
-        hasSubscore,
-        allAps,
-      );
+    if (transfer.courseName.startsWith('AP ')) {
+      const reorganized = organizeApExam(transfer as GroupedUncategorizedTransfer, toInsertAp, hasSubscore, allAps);
```
Comment (Awesome-E) [inline api/src/helpers/transferCredits.ts:365]: this cast seems like code smell, see if it's possible to eliminate?

File: site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx [hunk 8]
```diff
@@ -145,19 +153,70 @@ async function processTranscript(file: Blob) {
...
+        await trpc.transferCredits.overrideAllTransfers.mutate({
+          courses: mergedCourses,
+          ap: mergedAps,
+          ge: [],
+          other: mergedOther,
+        });
+      }
+
+      // Finally, set the actual plan
+      // This does not save automatically (the user has to save it manually)
```
Comment (Awesome-E) [inline site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx:219]: For consistency with z4p, we should make it save automatically right?

File: site/src/store/slices/transferCreditsSlice.ts [hunk 9]
```diff
@@ -1,17 +1,11 @@
 import { createSlice, PayloadAction } from '@reduxjs/toolkit';
-import { APExam, TransferredGE } from '@peterportal/types';
+import { APExam, TransferredGE, UserAPExam } from '@peterportal/types';
 
 export interface TransferredCourse {
   courseName: string;
   units: number;
 }
```
Comment (Awesome-E) [inline site/src/store/slices/transferCreditsSlice.ts:7]: I'd argue that these other types should also be refactored, and that the 3 should be named consistently

`TransferredCourse`, `TransferredAPExam`, `TransferredGE`, `TransferredUncategorizedItem` (idk last one is kinda weird)

and they should _all_ be in `@peterportal/types` for consistency. Even better, they should be consolidated with the types currently defined in the transfer credits controller by using `z.infer`

**Review by Awesome-E — CHANGES_REQUESTED** [review]

File: api/src/helpers/transferCredits.ts (hunk 1 above)
Comment (CadenLee2) [reply api/src/helpers/transferCredits.ts]: Fixed various comments in this file

File: api/src/helpers/transferCredits.ts (hunk 7 above)
Comment (CadenLee2) [reply api/src/helpers/transferCredits.ts:362]: Oh this was from the past when the null check was in this area, these casts are all unnecessary now and can be removed. Thanks for pointing it out

File: api/src/helpers/transferCredits.ts (hunk 4 above)
Comment (CadenLee2) [reply api/src/helpers/transferCredits.ts:326]: All these sections are now refactored into helper functions

File: site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx (hunk 8 above)
Comment (fiveminus1) [reply site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx:219]: It could be argued that since this saves to your current roadmap rather than a new one like Z4P does (which I personally think is fine), it shouldn't auto-save, but if that's the case then there should at least be a toast saying this isn't saved.

**Review by fiveminus1 — CHANGES_REQUESTED** [review]
Love the video! #qualitypr #cadenpeterportal

-- 

So, I can't test this with my own transcript because it's broken for some reason - will DM you it if you want to check it out, but it just leads to a blank page when I import it. This may or may not be directly related to this PR, since my transcript has never worked. 

I have tested my roommate's transcript, and the saving behavior locally/logged in seems to be working as expected (other than the auto-saving Ethan & I commented on). Will try to get some other sample transcripts and test some edge cases, then edit this review if I find anything

Edit: Testing with another transcript, and it appears AP Calculus AB isn't correctly categorized - it goes into "Other Transferred Credits."

**Comment by CadenLee2** [comment]
> So, I can't test this with my own transcript because it's broken for some reason

This should now be fixed: any invalid courses on the transcript will not be placed on the roadmap, and instead will also be included as other/uncategorized transfers. Duplicates should also be eliminated if there are any.

> it appears AP Calculus AB isn't correctly categorized

This is intentional, due to the fact that with certain transcript imports it could be ambiguous whether this is AB or AB Subscore on BC when we only have one instance of AP Calculus AB

File: api/src/controllers/transferCredits.ts [hunk 10]
```diff
@@ -120,7 +100,7 @@ const transferCreditsRouter = router({
       .where(eq(transferredGe.userId, ctx.session.userId!));
     return response as TransferredGE[];
   }),
-  setTransferredGE: userProcedure.input(z.object({ GE: zodGETransferSchema })).mutation(async ({ input, ctx }) => {
+  setTransferredGE: userProcedure.input(z.object({ GE: zodTransferredGE })).mutation(async ({ input, ctx }) => {
```
Comment (Awesome-E) [inline api/src/controllers/transferCredits.ts:103]: if you're updating the types, it'd probably be good to update this and make it more consistent as well.

```suggestion
  setTransferredGE: userProcedure.input(zodTransferredGE)).mutation(async ({ input, ctx }) => {
```

File: site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx (hunk 8 above)
Comment (Awesome-E) [reply site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx:245]: In my recent PR to consolidate things in the top navigation, it now adds to a new roadmap instead of overriding the current one

**Review by Awesome-E — APPROVED** [review]

**Review by fiveminus1 — APPROVED** [review]
Everything LGTM, good work and thanks for fixing my bug!!

File: api/src/controllers/transferCredits.ts (hunk 10 above)
Comment (CadenLee2) [reply api/src/controllers/transferCredits.ts]: Done

**Review by Awesome-E — APPROVED** [review]
